ports/esp32/machine_i2s.c

```diff
@@ -0,0 +1,141 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2017 Ayke van Laethem
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "machine_i2s.h"
+#include "py/runtime.h"
+#include "py/mperrno.h"
+
+STATIC const machine_i2s_obj_t machine_i2s_obj_0 = {
+    { &machine_i2s_type },
+    .port = I2S_NUM_0,
+};
+
+STATIC const machine_i2s_obj_t machine_i2s_obj_1 = {
+    { &machine_i2s_type },
+    .port = I2S_NUM_1,
+};
+
+STATIC void machine_i2s_obj_init_helper(const machine_i2s_obj_t *self, size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    enum { ARG_bclk, ARG_ws, ARG_data_out, ARG_data_in, ARG_samplerate, ARG_bitdepth };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_bclk, MP_ARG_REQUIRED | MP_ARG_INT },
+        { MP_QSTR_ws, MP_ARG_REQUIRED | MP_ARG_INT },
+        { MP_QSTR_data_out, MP_ARG_REQUIRED | MP_ARG_INT },
+        { MP_QSTR_data_in, MP_ARG_REQUIRED | MP_ARG_INT },
+        { MP_QSTR_samplerate, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 48000} },
+        { MP_QSTR_bitdepth, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 16} },
+    };
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    i2s_config_t config = {
+        .mode = I2S_MODE_MASTER | I2S_MODE_TX,
+        .sample_rate = args[ARG_samplerate].u_int,
+        .bits_per_sample = args[ARG_bitdepth].u_int,
+        .channel_format = I2S_CHANNEL_FMT_RIGHT_LEFT,
+        .communication_format = I2S_COMM_FORMAT_I2S | I2S_COMM_FORMAT_I2S_MSB,
+        .intr_alloc_flags = ESP_INTR_FLAG_LEVEL1,
+        .dma_buf_count = 8,
+        .dma_buf_len = 64,
```

Comment:
I've noticed this must be configurable somehow. For example, for stereo 48kHz 20ms packets (Opus) it needs to be at least 120.